p2p/memp2p/memp2p_network.go

```diff
@@ -0,0 +1,159 @@
+		peerIDs:     peerIDs,
+		peers:       make(map[p2p.PeerID]*MemP2PMessenger),
+		LogMessages: false,
+		Messages:    messages,
+	}
+}
+
+// ListAddresses provides the addresses of the known peers.
+func (network *MemP2PNetwork) ListAddresses() []string {
+	network.mutex.Lock()
+	addresses := make([]string, len(network.peerIDs))
+	i := 0
+	for _, peerID := range network.peerIDs {
+		addresses[i] = fmt.Sprintf("/memp2p/%s", peerID)
+		i++
+	}
+	network.mutex.Unlock()
+	return addresses
+}
+
+// ListAddressesExceptOne provides the addresses of the known peers, except a specified one.
+func (network *MemP2PNetwork) ListAddressesExceptOne(peerIDToExclude p2p.PeerID) []string {
+	network.mutex.Lock()
+	resultingLength := len(network.peerIDs) - 1
+	if resultingLength <= 0 {
+		network.mutex.Unlock()
+		return []string{}
+	}
+	addresses := make([]string, resultingLength)
+	k := 0
+	for _, peerID := range network.peerIDs {
+		if peerID == peerIDToExclude {
+			continue
+		}
+		addresses[k] = fmt.Sprintf("/memp2p/%s", peerID)
+		k++
+	}
+	network.mutex.Unlock()
+	return addresses
+}
+
+// Peers provides a copy of its internal map of peers
+func (network *MemP2PNetwork) Peers() map[p2p.PeerID]*MemP2PMessenger {
+	network.mutex.RLock()
+	peersCopy := make(map[p2p.PeerID]*MemP2PMessenger)
+	for peerID, peer := range network.peers {
+		peersCopy[peerID] = peer
+	}
+	network.mutex.RUnlock()
+	return peersCopy
+}
+
+// PeersExceptOne provides a copy of its internal map of peers, excluding a specific peer.
+func (network *MemP2PNetwork) PeersExceptOne(peerIDToExclude p2p.PeerID) map[p2p.PeerID]*MemP2PMessenger {
+	network.mutex.RLock()
+	peersCopy := make(map[p2p.PeerID]*MemP2PMessenger)
+	for peerID, peer := range network.peers {
+		if peerID == peerIDToExclude {
+			continue
+		}
+		peersCopy[peerID] = peer
+	}
+	network.mutex.RUnlock()
+	return peersCopy
+}
+
+// PeerIDs provides a copy of its internal slice of peerIDs
+func (network *MemP2PNetwork) PeerIDs() []p2p.PeerID {
+	network.mutex.RLock()
+	peerIDsCopy := make([]p2p.PeerID, len(network.peerIDs))
+	for i, peer := range network.peerIDs {
+		peerIDsCopy[i] = peer
+	}
+	network.mutex.RUnlock()
+	return peerIDsCopy
+}
+
+//PeerIDsExceptOne provides a copy of its internal slice of peerIDs, excluding a specific peer.
+func (network *MemP2PNetwork) PeerIDsExceptOne(peerIDToExclude p2p.PeerID) []p2p.PeerID {
+	network.mutex.RLock()
+	resultingLength := len(network.peerIDs) - 1
+	if resultingLength <= 0 {
+		network.mutex.RUnlock()
+		return []p2p.PeerID{}
+	}
+
+	peerIDsCopy := make([]p2p.PeerID, resultingLength)
+	k := 0
+	for _, peerID := range network.peerIDs {
+		if peerID == peerIDToExclude {
+			continue
+		}
+		peerIDsCopy[k] = peerID
+		k++
+	}
+	network.mutex.RUnlock()
+	return peerIDsCopy
+}
+
+// RegisterPeer adds a messenger to the Peers map and its PeerID to the peerIDs
+// slice.
+func (network *MemP2PNetwork) RegisterPeer(messenger *MemP2PMessenger) {
+	network.mutex.RLock()
```

Comment:
I guess this should be Lock/Unlock instead

Reply:
The function `Peers` could be called immediately after `RegisterPeer()` is called. Since `Peers` iterates directly over the `network.peers` map to read it, shouldn't we rather call `RLock` in `RegisterPeer()` to ensure that nobody reads the `network.peers` map while `RegisterPeer()` is writing it? I might be misunderstanding how `RLock`/`RUnlock` works though.

Reply:
I was misunderstanding. It must be `Lock`/`Unlock`, as you said.

Reply:
Fixed.